Add UpdateRequiredStatusChecks function

nmiyake opened the pull request:
Fixes #937

nmiyake commented on github/repos.go:
```diff
@@ -557,6 +557,12 @@ type RequiredStatusChecks struct {
 }
 
+// RequiredStatusChecksRequest represents a request to edit a protected branch's status checks.
+type RequiredStatusChecksRequest struct {
```
Adding request object because `UpdateRequiredStatusChecks` is a `PATCH` function so fields can be omitted in requests.

gmlewis commented on github/repos.go:
```diff
@@ -557,6 +557,12 @@ type RequiredStatusChecks struct {
+// RequiredStatusChecksRequest represents a request to edit a protected branch's status checks.
+type RequiredStatusChecksRequest struct {
+	Strict   *bool     `json:"strict"`
+	Contexts *[]string `json:"contexts"`
```
A slice is already a reference object, so this should be `[]string` and not `*[]string`.
Also, please add `omitempty` to this one as well.

nmiyake replied:
Will do for `omitempty`.

How important is consistency? I was trying to match the other code already in the repo, and other request objects with optional slices use pointers -- for example, in the same file:

```
type DismissalRestrictionsRequest struct {
	// The list of user logins who can dismiss pull request reviews. (Required; use nil to disable dismissal_restrictions or &[]string{} otherwise.)
	Users *[]string `json:"users,omitempty"`
	// The list of team slugs which can dismiss pull request reviews. (Required; use nil to disable dismissal_restrictions or &[]string{} otherwise.)
	Teams *[]string `json:"teams,omitempty"`
}
```

nmiyake replied:
Actually on second glance it looks like the file has both styles:

```go
type TransferRequest struct {
	NewOwner string  `json:"new_owner"`
	TeamID   []int64 `json:"team_id,omitempty"`
}
```

Will remove pointer reference.

gmlewis replied:
Consistency is very important in this repo, but in this case, the `*[]string` is not required I don't believe because an empty list here would just mean no contexts (if I'm reading the docs right). Do you agree?

nmiyake replied:
Yes, agree and updated in latest PR update.

gmlewis commented on github/repos.go:
```diff
@@ -557,6 +557,12 @@ type RequiredStatusChecks struct {
 
+// RequiredStatusChecksRequest represents a request to edit a protected branch's status checks.
+type RequiredStatusChecksRequest struct {
+	Strict   *bool     `json:"strict"`
```
Please add `omitempty` to both fields since they are optional.

gmlewis commented on github/repos.go:
```diff
@@ -791,6 +797,28 @@ func (s *RepositoriesService) RemoveBranchProtection(ctx context.Context, owner,
+// UpdateRequiredStatusChecks updates the required status checks for a given protected branch.
+//
+// GitHub API docs: https://developer.github.com/v3/repos/branches/#update-required-status-checks-of-protected-branch
+func (s *RepositoriesService) UpdateRequiredStatusChecks(ctx context.Context, owner, repo, branch string, sreq *RequiredStatusChecksRequest) (*RequiredStatusChecks, *Response, error) {
```
Even though the fields are optional, this call does nothing if `sreq` is `nil` so I'm thinking that we should either make `sreq` a value and not a pointer or add a check so that `nil` can not be passed.

nmiyake replied:
Should this behavior differ from other functions? I modeled the implementation of this after the existing `UpdateBranchProtection` function in the same file:

```
// UpdateBranchProtection updates the protection of a given branch.
//
// GitHub API docs: https://developer.github.com/v3/repos/branches/#update-branch-protection
func (s *RepositoriesService) UpdateBranchProtection(ctx context.Context, owner, repo, branch string, preq *ProtectionRequest) (*Protection, *Response, error) {
	u := fmt.Sprintf("repos/%v/%v/branches/%v/protection", owner, repo, branch)
	req, err := s.client.NewRequest("PUT", u, preq)
	if err != nil {
		return nil, nil, err
	}

	// TODO: remove custom Accept header when this API fully launches
	req.Header.Set("Accept", mediaTypeProtectedBranchesPreview)

	p := new(Protection)
	resp, err := s.client.Do(ctx, req, p)
	if err != nil {
		return nil, resp, err
	}

	return p, resp, nil
}
```

gmlewis replied:
Usually in the GitHub Developer API documentation, they mention the custom media headers on an endpoint-by-endpoint basis. So I'm thinking it doesn't apply here.

nmiyake replied:
Yes, agree and removed in latest PR update.

gmlewis replied:
Sorry for the delay, @nmiyake!
Hmmm... I see you are right about not checking `preq` for `nil`... seems odd to me but shows precedent, so let's run with it. We could always clean all these up later in a subsequent PR. This one has been held up long enough.

gmlewis commented on github/repos.go:
```diff
@@ -791,6 +797,28 @@ func (s *RepositoriesService) RemoveBranchProtection(ctx context.Context, owner,
+		return nil, nil, err
+	}
+
+	// TODO: remove custom Accept header when this API fully launches
```
nit: add period to end of sentence.

From the GitHub Developer API docs, it is not clear to me that this custom media type is required for this call.

nmiyake replied:
This is from the other function I used as a base for this one. However, agree that, based on the docs, this media type should not be required for this particular call -- will remove.

nmiyake commented:
Thanks for the review! Addressed most of the feedback -- will wait to hear back on whether I should change the function parameter or keep it consistent with the other ones.

nmiyake commented:
Ping -- still waiting on response to latest question (all other PR feedback has been addressed in update)

nmiyake commented:
Ping for another reviewer (@juliaferraioli or anyone else actively doing reviews) to unblock

gmlewis commented:
Thank you, @nmiyake and @juliaferraioli!
Merging.
